File: dynamic-programming/grid/minimum-falling-path-sum/ts/solution_state-compression-dp.ts
function minFallingPathSum(matrix: number[][]): number {
    const n = matrix[0].length
    const current: number[] = Array(n + 2).fill(0)
    current[0] = current[n + 1] = Number.MAX_SAFE_INTEGER
    for (const row of matrix) {
        let prev: number = current[0]
        for (let j = 0; j < n; j++) {
            const temp = prev
            prev = current[j + 1]
            current[j + 1] = Math.min(temp, current[j + 1], current[j + 2]) + row[j]
        }
    }
    let minSum: number = Number.MAX_SAFE_INTEGER
    for (let i = 1; i < n + 1; i++) {
        minSum = Math.min(minSum, current[i])
    }
    return minSum
}

export { minFallingPathSum }

File: dynamic-programming/grid/minimum-falling-path-sum/java/Solution_StateCompressionDp.java
class Solution_StateCompressionDp {
    public int minFallingPathSum(int[][] matrix) {
        int n = matrix[0].length;
        int[] current = new int[n + 2];
        current[0] = current[n + 1] = Integer.MAX_VALUE;
        for (int[] row : matrix) {
            int prev = current[0];
            for (int j = 0; j < n; j++) {
                int temp = prev;
                prev = current[j + 1];
                current[j + 1] = Math.min(Math.min(temp, current[j + 1]), current[j + 2]) + row[j];
            }
        }
        int minSum = Integer.MAX_VALUE;
        for (int i = 1; i < n + 1; i++) {
            minSum = Math.min(minSum, current[i]);
        }
        return minSum;
    }
}